src/nfa/nfa.cc

```diff
@@ -312,6 +312,136 @@ struct StackLevel {
     };
 };
 
+struct NodeData {
+    Post::const_iterator post_it;
+    Post::const_iterator post_end;
+    StateSet::const_iterator targets_it{};
+    StateSet::const_iterator targets_end{};
+    unsigned long index;
+    unsigned long lowlink;
+    bool initilized;
+    bool on_stack;
+
+    NodeData() = default;
+
+    NodeData(State q, const Delta & delta, unsigned long index) 
+        : post_it(delta[q].cbegin()), post_end(delta[q].cend()), index(index), lowlink(index), initilized(true), on_stack(true) {
+        if (post_it != post_end) {
+            targets_it = post_it->cbegin();
+            targets_end = post_it->cend();
+        }
+    };
+
+    // TODO: this sucks. In fact, if you want to check that you have the last sucessor, you need to 
+    // first align the iterators.
+    // TODO: this is super-ugly. If we introduce Post::transitions iterator, this could be much easier.
```

Comment:
It seems that we agreed to create the iterator over transitions. How will it improve the code here if we have to allocate each new transition `Trans` every iteration?

Reply:
The code will be simpler.

Reply:
That is true. But it might be potentially slower, right? I do not understand how you and Lukáš intend to iterate over triples without allocating them.

Reply:
Yes. But if the difference will be negligible, I suggest to use the transition iteration in order to keep the code at least a bit readable.

Reply:
Definitely. It would be so much more readable, indeed. I consider the conversation resolved. When we implement the transition iterator, we can think about this again.